message/consumer.go

```diff
@@ -6,92 +6,94 @@ import (
+func CloseConsumer() {
+	closer <- true
+}
+
+func Consume(in KafkaConsumer, out CompletedProducer, eventHandler EventHandler, errorEventHandler ErrorEventHandler) {
 	go func() {
-		for {
+		consuming := true
+		for consuming {
 			select {
-			case consumedMessage := <-dimensionExtractedConsumer.Incoming():
+			case consumedMessage := <-in.Incoming():
 				consumedData := consumedMessage.GetData()
-				var dimensionsExtractedEvent event.DimensionsExtractedEvent
-				if err := schema.DimensionsExtractedSchema.Unmarshal(consumedData, &dimensionsExtractedEvent); err != nil {
-					exitChannel <- err
-					return
+				var newInstanceEvent event.NewInstanceEvent
+				if err := schema.NewInstanceSchema.Unmarshal(consumedData, &newInstanceEvent); err != nil {
+					log.ErrorC(unmarshallErrMsg, err, nil)
+					errorEventHandler.Handle("", err, nil)
+					consumedMessage.Commit()
+					continue
 				}
-				logData := map[string]interface{}{eventKey: dimensionsExtractedEvent}
+
+				logData := map[string]interface{}{eventKey: newInstanceEvent}
 				log.Debug(eventRecieved, logData)
 
-				if err := eventHandler.HandleEvent(dimensionsExtractedEvent); err != nil {
+				if err := eventHandler.HandleEvent(newInstanceEvent); err != nil {
 					log.ErrorC(eventHandlerErr, err, logData)
-					exitChannel <- err
-					return
+					errorEventHandler.Handle(newInstanceEvent.InstanceID, err, nil)
+					consumedMessage.Commit()
+					continue
 				}
 
-				logData[logKeys.InstanceID] = dimensionsExtractedEvent.InstanceID
+				logData[logKeys.InstanceID] = newInstanceEvent.InstanceID
 				log.Debug(eventHandlerSuccess, logData)
 
-				insertedEvent := event.DimensionsInsertedEvent{
-					FileURL:    dimensionsExtractedEvent.FileURL,
-					InstanceID: dimensionsExtractedEvent.InstanceID,
+				insertedEvent := event.InstanceCompletedEvent{
+					FileURL:    newInstanceEvent.FileURL,
+					InstanceID: newInstanceEvent.InstanceID,
 				}
 
-				if err := insertedProducer.DimensionInserted(insertedEvent); err != nil {
-					exitChannel <- err
-					return
+				if err := out.Completed(insertedEvent); err != nil {
+					errorEventHandler.Handle(newInstanceEvent.InstanceID, err, nil)
+					consumedMessage.Commit()
+					continue
 				}
 				consumedMessage.Commit()
-
-			case consumerError := <-dimensionExtractedConsumer.Errors():
-				log.ErrorC(conumserErrMsg, consumerError, log.Data{logKeys.ErrorDetails: consumerError})
-				dimensionExtractedConsumer.Closer() <- true
-				insertedProducer.Closer() <- true
-				exitChannel <- consumerError
-				return
-			case producerError := <-insertedProducer.Errors():
-				log.ErrorC(producerErrMsg, producerError, nil)
-				dimensionExtractedConsumer.Closer() <- true
-				insertedProducer.Closer() <- true
-				exitChannel <- producerError
-				return
+			case <-closer:
+				log.Info(consumerStoppedMsg, nil)
+				consuming = false
```

Comment:
Consider replacing this with a return statement. All that happens when consuming = false is the for loop breaks. You can take the condition out of the for loop and let it run forever, and then just return when you get a signal on  the closer channel.

Reply:
Done - consume now take a context to determined when to close.